File: timer.py
import sys
from Arc import Arc, EnumTimingSense
import CircuitBuilder
from Pin import Pin
from EnumClass import ALL_CLOCK_EDGES, ALL_TIMING_MODES, FOREACH_EL_FRF_TRF, FOREACH_EL_RF, EnumClockEdge, EnumPinType, EnumTimingMode

def toposort(circuit: CircuitBuilder):
    visited = set()
    stack = []

    def dfs(pin: Pin):
        if pin in visited:
            return
        visited.add(pin)
        for arc in pin.fanout:
            dfs(arc.to_pin)
        stack.append(pin)

    for pin in circuit.primary_inputs.values():
        dfs(pin)

    return stack[::-1]  # Reverse the stack to get the correct order

def get_all_paths(start_pin: Pin, path):
    if len(start_pin.fanout) == 0:
        yield path.copy()
        return
    for arc in start_pin.fanout:
        assert isinstance(arc, Arc), f"Arc {arc} is not an instance of Arc"
        path.append(arc)
        yield from get_all_paths(arc.to_pin, path)
        path.pop()

class Timer:
    def __init__(self, circuit: CircuitBuilder, clock_cycle= 10, clock_rise_at = None, clock_fall_at = None):
        self.circuit: CircuitBuilder = circuit
        self.clock_pin = None
        self.clock_cycle = clock_cycle
        self.clock_rise_at = 0 if clock_rise_at is None else clock_rise_at
        self.clock_fall_at = clock_cycle /2 + self.clock_rise_at if clock_fall_at is None else clock_fall_at

    def update_capacitance(self):
        for pin in self.circuit.pin_factory.get_all_pins():
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            if pin.type in (EnumPinType.PRIMARY_OUTPUT, EnumPinType.INPUT):
                continue
            FOREACH_EL_RF(pin.update_capacitance)

    def propagate_slew(self):
        # 拓扑排序所有Pin, 传播slew
        for pin in toposort(self.circuit):
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            FOREACH_EL_FRF_TRF(pin.propagate_slew)
    
    def propagate_delay(self):
        # 拓扑排序所有Pin, 传播delay
        for pin in toposort(self.circuit):
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            FOREACH_EL_FRF_TRF(pin.propagate_delay)
    
    def propagate_arrival_time(self):
        # 拓扑排序所有Pin, 传播arrival_time
        for pin in toposort(self.circuit):
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            FOREACH_EL_FRF_TRF(pin.propagate_arrival_time)

    def propagate_request_arrival_time(self):
        for pin in self.circuit.primary_outputs.values():
            def reset_rat(el, rf):
                pin.set_request_arrival_time(el, rf, self.clock_cycle + self.clock_rise_at)
            FOREACH_EL_RF(reset_rat)
        
        for arc in self.circuit.get_all_constraint_arcs():
            assert isinstance(arc, Arc), f"Arc {arc} is not an instance of Arc"
            FOREACH_EL_FRF_TRF(arc.propagate_request_arrival_time)

    def reset_arrival_time(self):
        for pin in self.circuit.pin_factory.get_all_pins():
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            if pin.type == EnumPinType.PRIMARY_INPUT:
                pin.set_arrival_time(EnumTimingMode.MAX, EnumClockEdge.RISING, 0)
                pin.set_arrival_time(EnumTimingMode.MAX, EnumClockEdge.FALLING, 0)
                continue
            def reset_at(el, rf):
                pin.set_arrival_time(el, rf, None)
                pin.set_arrival_time(el, rf, None)
            FOREACH_EL_RF(reset_at)
    
    def deduce_clock(self):
        if self.clock_pin is not None:
            return self.clock_pin
        # 推断时钟引脚
        for cell in self.circuit.cells:
            if "CK" in cell.port_mapping:
                clock_pin_name = cell.port_mapping["CK"]
                clock_pin = self.circuit.get_pin(clock_pin_name)
                if clock_pin and clock_pin.type == EnumPinType.PRIMARY_INPUT:
                    self.clock_pin = clock_pin
                    return clock_pin
        return None
    
    def report_timing(self, delay_type: str, start_type):
        if self.clock_pin is None:
            print("Warning: Clock pin not found, Trying to deduce clock pin...", file=sys.stderr)
            self.clock_pin = self.deduce_clock()

        paths = {
            "in2reg": [],
            "in2out": [],
            "reg2reg": [],
            "reg2out": []
        }
        self.reset_arrival_time()
        # 拓扑排序所有Pin, 传播arrival_time
        for pin in toposort(self.circuit):
            assert isinstance(pin, Pin), f"Pin {pin} is not an instance of Pin"
            if start_type == "in":
                if pin == self.clock_pin:
                    continue
            if start_type == "reg":
                assert self.clock_pin, "Cannot find clock pin for reg2reg or reg2out path"
                if pin in self.circuit.primary_inputs.values() and pin != self.clock_pin:
                    continue
            FOREACH_EL_FRF_TRF(pin.propagate_arrival_time)
        el = EnumTimingMode.to_enum(delay_type)
        # 从所有没有fanout的Pin开始回溯, 生成路径
        endpoints = [pin for pin in self.circuit.pin_factory.get_all_pins() if len(pin.fanout) == 0]
        for end_pin in endpoints:
            if end_pin.type == EnumPinType.PRIMARY_OUTPUT:
                end_type = "out"
            elif end_pin.type == EnumPinType.INPUT:
                end_type = "reg"
            else:
                continue
            pin = end_pin
            assert isinstance(end_pin, Pin), f"Pin {end_pin} is not an instance of Pin"
            atr = pin.arrival_time[el][EnumClockEdge.RISING]
            atf = pin.arrival_time[el][EnumClockEdge.FALLING]
            ratr = pin.request_arrival_time[el][EnumClockEdge.RISING]
            ratf = pin.request_arrival_time[el][EnumClockEdge.FALLING]
            slackr = ratr - atr if atr is not None and ratr is not None else None
            slackf = ratf - atf if atf is not None and ratf is not None else None
            if slackf is None or slackr is None:
                continue
            if slackr > slackf:
                edge = EnumClockEdge.RISING
            else:
                edge = EnumClockEdge.FALLING
            # if delay_type == "max":
            #     if atr > atf:
            #         edge = EnumClockEdge.RISING
            #     else:
            #         edge = EnumClockEdge.FALLING
            # else:
            #     if atr < atf:
            #         edge = EnumClockEdge.RISING
            #     else:
            #         edge = EnumClockEdge.FALLING
            path = []
            while True:
                path.append((pin.name , edge , pin.arrival_time[el][edge]))
                if pin.predecessor[el][edge] is None:
                    paths[f"{start_type}2{end_type}"].append(path[::-1])
                    break
                edge, arc = pin.predecessor[el][edge]
                pin = arc.from_pin
        return paths
                

            
    

File: src/Timer.cc
//
// Created by wenz on 2/26/26.
//

#include "Timer.h"

#include "Enum/EnumForeach.h"

namespace mySTA {
namespace {

bool should_skip_start_pin(const Pin* pin, const Pin* clock_pin, const EnumPointType start_type)
{
  if (start_type == EnumPointType::IN && pin == clock_pin) {
    return true;
  }
  if (start_type == EnumPointType::REG && pin->is_primary_input() && pin != clock_pin) {
    return true;
  }
  return false;
}

bool is_matching_end_point(const Pin* pin, const EnumPointType end_type)
{
  if (end_type == EnumPointType::OUT) {
    return pin->is_primary_output();
  }
  if (end_type == EnumPointType::REG) {
    return pin->is_input();
  }
  return false;
}

}  // namespace

Timer::Timer(Circuit& circuit, std::optional<float_t> clock_cycle, std::optional<float_t> clock_rise_at,
             std::optional<float_t> clock_fall_at)
    : _circuit(circuit),
      _clock_cycle(clock_cycle.value_or(_default_clock_cyle)),
      _clock_rise_at(clock_rise_at.value_or(_default_clock_rise_at)),
      _clock_fall_at(clock_fall_at.value_or(_default_clock_fall_at(_clock_cycle, _clock_rise_at)))
{
}

void Timer::update_capacitance()
{
  LOG_INFO << "update capacitance start";
  for (const auto& pins{_circuit.get_toposorted_pins()}; auto* pin : pins) {
    LOG_ASSERT(dynamic_cast<Pin*>(pin) != nullptr) << std::format("Pin is not an instance of Pin");
    const auto pin_type{pin->get_type()};
    if (pin_type == EnumPinType::PRIMARY_OUTPUT || pin_type == EnumPinType::INPUT) {
      continue;
    }
    FOREACH_EL_RF([pin](const EnumTimingMode el, const EnumClockEdge rf) { pin->update_capacitance(el, rf); });
  }
  LOG_INFO << "update capacitance end";
}

void Timer::propagate_slew()
{
  LOG_INFO << "propagate slew start";
  for (const auto& pins{_circuit.get_toposorted_pins()}; auto* pin : pins) {
    FOREACH_EL_FRF_TRF([pin](const auto el, const auto frf, const auto trf) { pin->propagate_slew(el, frf, trf); });
  }
  LOG_INFO << "propagate slew end";
}

void Timer::propagate_delay()
{
  LOG_INFO << "propagate delay start";
  for (const auto& pins{_circuit.get_toposorted_pins()}; auto* pin : pins) {
    FOREACH_EL_FRF_TRF([pin](const auto el, const auto frf, const auto trf) { pin->propagate_delay(el, frf, trf); });
  }
  LOG_INFO << "propagate delay end";
}

void Timer::propagate_arrival_time()
{
  LOG_INFO << "propagate arrival time start";
  for (const auto& pins{_circuit.get_toposorted_pins()}; auto* pin : pins) {
    FOREACH_EL_FRF_TRF([pin](const auto el, const auto frf, const auto trf) { pin->propagate_arrival_time(el, frf, trf); });
  }
  LOG_INFO << "propagate arrival time end";
}

void Timer::propagate_request_arrival_time()
{
  LOG_INFO << "propagate request arrival time start";
  for (const auto& primary_outputs{_circuit.get_primary_outputs()}; auto* pin : primary_outputs) {
    FOREACH_EL_RF(
        [this, pin](EnumTimingMode el, EnumClockEdge rf) { pin->set_request_arrival_time(el, rf, _clock_cycle + _clock_rise_at); });
  }

  for (const auto& constraint_arcs{_circuit.get_constraint_arcs()}; auto* arc : constraint_arcs) {
    FOREACH_EL_FRF_TRF(
        [arc](EnumTimingMode el, EnumClockEdge frf, EnumClockEdge trf) { arc->propagate_request_arrival_time(el, frf, trf); });
  }
  LOG_INFO << "propagate request arrival time end";
}

void Timer::reset_arrival_time()
{
  for (const auto& pins{_circuit.get_all_pins()}; auto* pin : pins) {
    pin->reset_all_arrival_time();
  }
}

void Timer::reset_request_arrival_time()
{
  for (const auto& pins{_circuit.get_all_pins()}; auto* pin: pins) {
    pin->reset_all_request_arrival_time();
  }
}

const Pin* Timer::deduce_clock()
{
  if (_clock_pin == nullptr) {
    _clock_pin = _circuit.deduce_clock();
  }
  return _clock_pin;
}

void Timer::build_report_cache(const EnumPointType start_type)
{
  if (_report_cache[+start_type][+EnumTimingMode::MAX][+EnumPointType::REG].has_value()) {
    return;
  }

  reset_arrival_time();

  for (const auto& pins{_circuit.get_toposorted_pins()}; auto* pin : pins) {
    if (should_skip_start_pin(pin, _clock_pin, start_type)) {
      continue;
    }
    FOREACH_EL_FRF_TRF([&](const auto el, const auto frf, const auto trf) { pin->propagate_arrival_time(el, frf, trf); });
  }

  const std::vector end_points{_circuit.get_all_pins() | std::views::filter([](auto* pin) { return pin->get_fanout().empty(); })
                               | std::ranges::to<std::vector<Pin*>>()};

  const auto collect_paths = [&](const EnumTimingMode timing_mode, const EnumPointType end_type) {
    report_paths_t paths{};
    for (const auto* end_point : end_points) {
      if (!is_matching_end_point(end_point, end_type)) {
        continue;
      }
      const auto* pin{end_point};
      auto atr{pin->get_arrival_time(timing_mode, EnumClockEdge::RISING)};
      auto atf{pin->get_arrival_time(timing_mode, EnumClockEdge::FALLING)};
      auto ratr{pin->get_request_arrival_time(timing_mode, EnumClockEdge::RISING)};
      auto ratf{pin->get_request_arrival_time(timing_mode, EnumClockEdge::FALLING)};
      if (!atr or !atf or !ratr or !ratf) {
        continue;
      }
      float_t slack_r{};
      float_t slack_f{};
      switch (timing_mode) {
        case EnumTimingMode::MAX:
          slack_r = _clock_cycle - *ratr - *atr;
          slack_f = _clock_cycle - *ratf - *atf;
          break;
        case EnumTimingMode::MIN:
          slack_r = *atr - *ratr;
          slack_f = *atf - *ratf;
          break;
        default:
          break;
      }
      float_t slack{};
      EnumClockEdge edge{};
      if (slack_r < slack_f) {
        slack = slack_r;
        edge = EnumClockEdge::RISING;
      } else {
        slack = slack_f;
        edge = EnumClockEdge::FALLING;
      }
      std::vector<path_t> path{};
      while (true) {
        const auto arrival_time{pin->get_arrival_time(timing_mode, edge)};
        const auto slew{pin->get_slew(timing_mode, edge)};
        LOG_ASSERT(arrival_time);
        LOG_ASSERT(slew);
        path.emplace_back(std::string{pin->get_name()}, edge, *arrival_time, slack, pin->get_capacitance(timing_mode, edge), *slew);
        const auto predecessor{pin->get_predecessor(timing_mode, edge)};
        if (!predecessor) {
          std::ranges::reverse(path);
          paths.push_back(path);
          break;
        }
        edge = predecessor->first;
        pin = predecessor->second->from_pin();
      }
    }
    return paths;
  };

  for (const auto timing_mode : {EnumTimingMode::MAX, EnumTimingMode::MIN}) {
    for (const auto end_type : {EnumPointType::REG, EnumPointType::OUT}) {
      _report_cache[+start_type][+timing_mode][+end_type] = collect_paths(timing_mode, end_type);
    }
  }
}

const Timer::report_paths_t& Timer::report_timing(const EnumTimingMode timing_mode, const EnumPointType start_type, const EnumPointType end_type)
{
  LOG_ASSERT(start_type != EnumPointType::OUT);
  LOG_ASSERT(end_type != EnumPointType::IN);
  LOG_ASSERT(timing_mode == EnumTimingMode::MAX || timing_mode == EnumTimingMode::MIN);
  if (_clock_pin == nullptr) {
    LOG_WARNING << std::format("Warning: Clock not found, Trying to deduce clock pin...");
    deduce_clock();
  }
  build_report_cache(start_type);
  LOG_ASSERT(_report_cache[+start_type][+timing_mode][+end_type]);
  return _report_cache[+start_type][+timing_mode][+end_type].value();
}

}  // namespace mySTA
